### DISCUSSION TITLE
Adding and modifying Tiled layers?

### DISCUSSION THREAD

**Discussion by DaVince** [excaliburjs/Excalibur#1894] [Q&A]
I'm using [Excalibur-Tiled](https://github.com/excaliburjs/excalibur-tiled/issues) in my game to load maps. It has been doing its job pretty decently, though I miss a few features here and there (or at least I don't know where to look if I want to use them, like identifying a cell by its name or type rather than its numerical ID in the tilesheet).

Anyway, the above is unrelated. What I'm trying to do is:
* Dynamically add a layer on top of the actual layers in the Tiled map;
* Fill that layer with a specific tile;
* Modify certain tiles by their x/y coordinates on the fly

My aim is basically to have a "fog of war" effect and leverage the existing tilemap to do it. But I'm kind of running into a roadblock:
* The TiledResource has a layers property with an array of layers. It looks I can push a new layer on top of it, but that means entering actual Tiled data manually. I was hoping there was a createLayer() convenience function or something that at least sets it up. And when I do, I don't think changing this data on the fly will have an effect on the map unless I rerun and re-add getTileMap(). That'll get rid of whatever tiles on the other layers that I'm changing on the fly, though.
* Excalibur's own TileMap type doesn't seem to have the concept of layers; just one huge array of all cells and their properties, and I don't know how z-indexing is done on those cells.

Any tips or methods on how to tackle this? Maybe excalibur-tiled could use some more love? :slightly_smiling_face:

**Comment by eonarheim** [comment]
You are definitely right, excalibur-tiled needs an update 😃 I have a branch of excalibur-tiled going that might be close to what you need (hoping to merge soon with the next release of excalibur which will happen as soon as I've finished the [collision update PR](https://github.com/excaliburjs/Excalibur/pull/1818))

Excalibur-Tiled PR https://github.com/excaliburjs/excalibur-tiled/pull/162 (theoretically compatible with the current released version of excalibur if you want to try running it, but to be honest is probably best on latest main)

I've added a lot of missing features that folks have been wanting, like working with multiple Tiled layers and Tiled objects with arbitrary data.

What do you think of the documented methods in the new proposed readme? Definitely open to more features or tweaks
* https://github.com/excaliburjs/excalibur-tiled/blob/7164e3a29704bbe70e574248380187014c05cdf7/README.md

To accomplish the fog of war effect, it might be possible to use another TileMap positioned above the current level, then use semi-transparent/opaque graphics and turn them on and off. This should work without re-adding the TileMap to the scene. 

We also did a somewhat similar thing to what you're trying to do for LD48 to make diggable dirt given x/y of the cell https://github.com/excaliburjs/ludum-48/blob/main/src/level.ts#L127. Also we used z-indexed TileMaps as backgrounds https://github.com/excaliburjs/ludum-48/blob/8d00d06b3d637f064f7b325429ff7d073958375c/src/background.ts#L6). Maybe you'll be able to repurpose some of this code? 

Individual cells can be layered by adding graphics in painter order, or individual TileMaps (at least on latest main) can be positioned with z-index to achieve layering in the scene.

**Reply by DaVince** [reply]
Wow, excalibur-tiled has been given a lot of love and attention! All of it looks like it's an improvement; I'll try it out in a branch and see how much of the cool new stuff I can use and what I feel might still be missing. :smiley: 

Thank you for the suggestions and the the resources, too. I'll look at the code and try the idea of just separating the fog into its own TileMap. Much better idea, really.

Edit: I did just spot two instances of a typo: `TiledMapeResource`.
